keps/sig-cli/2590-kubectl-subresource/README.md

```diff
@@ -0,0 +1,386 @@
+```
+
+## Design Details
+
+### Subresource support
+
+A new field `Subresource` and method `Subresource`/`WithSubresource` is added to
+the [builder], [helper] and [visitor] code to use the API object at the subresource path.
+
+[builder]: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cli-runtime/pkg/resource/builder.go
+[helper]: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cli-runtime/pkg/resource/helper.go
+[visitor]: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cli-runtime/pkg/resource/visitor.go
+
+For each relevant kubectl command, a new flag `--subresource` is added. If the flag is specified,
+the value is also validated to ensure that it is a valid subresource.
+
+If the subresource does not exist for an API resource, a `NotFound` error is returned.
+
+### Table printer
+
+To support table view for subresources using kubectl get, table convertor support is added to
+the scale and status subresoruces for built-in and CRD types.
+
+For built-in types, `StatusStore` and `ScaleStore` are updated to implement the `TableConvertor` interface.
+`StatusStore` uses the same columns as the main resource object.
+
+The following column definitions for the `Scale` object are added to [printers.go] to support the scale subresource:
+- `Available Replicas` uses the json path `.status.replicas` of autoscalingv1.Scale object
+- `Desired Replicas` uses the json path of `.spec.replicas` of autoscalingv1.Scale object
+
+For custom resources:
+- the status subresoruce uses the same columns as defined for the full resource, i.e., `additionalPrinterColumns` defined in the CRD.
+- the scale subresource follows the same column definitions as the built-in types, and are defined in [helpers.go].
+
+[printers.go]: https://github.com/kubernetes/kubernetes/blob/master/pkg/printers/internalversion/printers.go#L88
+[helpers.go]: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/helpers.go
+
+### Test Plan
+
+- Unit tests, integration and e2e tests will be added.
+
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+
+- [ ] Collect user feedback on adding support of `--subresource` for `apply`
+
+#### Beta -> GA Graduation
+
+- [ ] User feedback gathered for atleast 1 cycle
+
+### Upgrade / Downgrade Strategy
+
+This functionality is contained entirely within kubectl and shares its strategy.
+No configuration changes are required.
+
+### Version Skew Strategy
+
+Not applicable. There is nothing required of the API Server, so there
+can be no version skew.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
+-->
+
+- [ ] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name:
+    - Components depending on the feature gate:
+- [x] Other
```

Comment:
We don't have an explicit mechanism for stating if a flag is alpha or not, there are a couple of approaches:
1. Use `alpha` prefix in the flag name, but I personally don't like that approach since it means we will have to change flag name eventually.
2. Force using env var, which allows this flag.
3. Hiding the flag by default, similarly to how deprecated flags are hidden.
4. Providing a description which is explicitly calling out it's alpha.

I'm personally leaning towards 3 or 4, and given this is an explicit opt-in (you need to place the flag when invoking kubectl command) I'm not worried that users will have issues with that.

Reply:
+1 for having a description explicitly calling out it's alpha.

Reply:
Ok, can you pick whatever mechanism is preferred by the SIG and document it here. I don't have a preference really. Otherwise all looks good for PRR, I will approve after that change and the SIG approves.

Reply:
Updated to option 4 (description explicitly calling out as alpha) and squashed the commits. PTAL :)